### ISSUE TITLE
Problem using latest version (0.6.6)

### ISSUE THREAD

**Issue by PezzoniL** [microsoft/vscode-iot-workbench#849]
I had updated the azure SDK to the latest public_preview.
And I had the necessity to regenerate the interfaces because they are not compatible with the latest version..

So I want to use your fantastic visual studio plugin..
Visual code has downloaded the latest plugin version (v0.6.6)

I’m able to open public/company repository… 

But if I want to “Generate skeleton device code and project based on given device capability model file”…

I had:

**[IoT Plug and Play] Check IoT Plug and Play CodeGen Cli ... v0.6.6 is installed and ready to use.

Unhandled Exception: 
System.AggregateException: One or more errors occurred. (No such host is known) ---> System.Net.Http.HttpRequestException: No such host is known ---> System.Net.Sockets.SocketException: No such host is known
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Azure.IoT.DigitalTwin.CodeGen.Program.GetCliLatestVersion() in D:\a\1\s\src\Azure.IoT.DigitalTwin.CodeGen.Cli\Program.cs:line 82
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at Azure.IoT.DigitalTwin.CodeGen.Program.NofifyIfUpdateAvailable() in D:\a\1\s\src\Azure.IoT.DigitalTwin.CodeGen.Cli\Program.cs:line 90
   at Azure.IoT.DigitalTwin.CodeGen.Program.Main(String[] args) in D:\a\1\s\src\Azure.IoT.DigitalTwin.CodeGen.Cli\Program.cs:line 30

[IoT Plug and Play] generate PnP device code failed.**

Probably we need to put under our company proxy a further exception for a new url that you want to use…
But which url do we need to add?

Thanks a lot
luca

**Comment by dooriya** [comment]
@PezzoniL Thank you for the feedback.
According to the information you provided, seems you failed to get the latest codegen version info from the URL  https://aka.ms/codegen-latest-version.

Are you able to access this URL within your company network?

**Comment by PezzoniL** [comment]
Yes. I'm able to reach it..
it reports the latest version:
0.6.6

also because.. looking the CodeGeneratorCore.ts in function:
generateDeviceCodeStub

the sequence is:
`// Check installation of Codegen CLI and update its bits if a new version is`
    
**I'm able to reach it**
`    // Select Capability Model`
    
**I can select my Model**
```
    // Prompt if old project exists for the same Capability Model file
   
    // Specify project name
```
    
**I can specify the project name**

`    // Select language`
  
**I can select AnsiC**
`    // Read CodeGen options configuration JSON`
  
`    // Select device connection string type`
**I can select iot hub device connection string**
  

`    // Select project template`
  **I can select CMAke project**

`    // Select Device SDK reference type for CMake projec`t
  **I can select "via source code"**

**So the problem is after this point...**




  ```
  // Download dependent interface of capability model

    if (!await DigitalTwinUtility.downloadDependentInterface(
            rootPath, capabilityModelFilePath)) {
      return;
    }

    const codeGenExecutionInfo: CodeGenExecutionItem = {
      outputDirectory: path.join(rootPath, codeGenProjectName),
      capabilityModelFilePath,
      interfaceDirecoty: rootPath,
      projectName: codeGenProjectName,
      languageLabel: CodeGenLanguage.ANSIC,
      codeGenProjectType,
      deviceSdkReferenceType: sdkReferenceType,
      deviceConnectionType: connectionType
    };

    this.saveCodeGenConfig(
        rootPath, capabilityModelFilePath, codeGenExecutionInfo);

    await this.generateDeviceCodeCore(
        codeGenExecutionInfo, context, channel, telemetryContext);
  }
```
And inside Visual Studio I'm able to open both company and public model repositories.

**Comment by PezzoniL** [comment]
looking the content of .codeGenConfigs file:
```
{
    "codeGenExecutionItems": [
        {
            "outputDirectory": "c:\\nobackup\\azure-iot-sdk-c\\STWinNew14\\stwin",
            "capabilityModelFilePath": "c:\\nobackup\\azure-iot-sdk-c\\STWinNew14\\stwin.capabilitymodel.json",
            "interfaceDirecoty": "c:\\nobackup\\azure-iot-sdk-c\\STWinNew14",
            "projectName": "stwin",
            "languageLabel": "ANSI C",
            "codeGenProjectType": "CMake_Windows",
            "deviceSdkReferenceType": "SourceCode",
            "deviceConnectionType": "ConnectionString"
        }
    ]
}
```

I think that I'm able also to reach:
```
this.saveCodeGenConfig(
      rootPath, capabilityModelFilePath, codeGenExecutionInfo);
```
so my problem is located in this.generateDeviceCodeCore
thanks a lot

**Comment by dooriya** [comment]
Hi @PezzoniL, 

The exception is not thrown in the Workbench Extension, it is thrown from the CodeGen CLI tool, which source code is not open-source yet.

This is the code in **CodeGen CLI** (not the Workbench Extension) where the exception is thrown.
```csharp
private static async Task<string> GetCliLatestVersion()
{
	const string versionInfo = "https://aka.ms/codegen-latest-version";

	using (HttpClient httpClient = new HttpClient())
	{
		var response = await httpClient.GetAsync(versionInfo);
		response.EnsureSuccessStatusCode();
		return await response.Content.ReadAsStringAsync();
	}
}
```
According to the exception message you provided:

> System.AggregateException: One or more errors occurred. (No such host is known) ---> System.Net.Http.HttpRequestException: No such host is known ---> System.Net.Sockets.SocketException: No such host is known
at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)

The `HttpRequestException` you got is due to the underlying **DNS failure** (see [GetAsync API ref](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.getasync?view=netcore-2.2)). 

This issue occurs because the agent computer (your PC) sends its Network Basic Input/Output System (NetBIOS) name instead of sending its Fully Qualified Domain Name (FQDN) name to the server. When the DNS server of the service does not have the IP address mapping of the NetBIOS name of the agent computer, the issue occurs.

So could you please check your local DNS server to make sure it's configured correctly?  And try a workaround to update your local DNS server to map the IP address to the NetBIOS name of your computer? or update your company proxy to add the [target URL](https://aka.ms/codegen-latest-version)?

**Comment by dooriya** [comment]
At the same time, we will update the CodeGen CLI to **ignore** the `HttpRequestException` thrown from the version check. The code generation should not be blocked by the latest version check failure here.

**Comment by PezzoniL** [comment]
I had tested and we are able to reach the https://aka.ms/codegen-latest-version
BUT if I use one browser that it's able to speak with the proxy.
Speaking with our IT team (I work for STMicrolectronics)  they said to me that if you need to implement the capability to speak with one proxy (the previous version of this plugin was working without any problem for us)
Otherwise this plugin could not work with (for what our IT team said) with **any** company proxy...

I will send one email to using dol@microsoft.com from luca.pezzoni@st.com putting in cc also our member of our IT team that will be able to explain the problem in a better way respect me.
Thanks again for all your help
Ciao
Luca

**Comment by dooriya** [comment]
Hi @PezzoniL,
 
We've just released a new CodeGen version (**0.6.7**) to improve the logic of the version check.
Could you please re-run the codegen command (the **0.6.7** version of the CodeGen will be auto-upgraded) from VS Code command palette to see if this new release works on your environment?

**Comment by PezzoniL** [comment]
It works!
Thanks a lot for your very fast implementation!
